types/yup/index.d.ts

```diff
@@ -170,16 +171,18 @@ export interface ArraySchema<T> extends Schema<T[]> {
     compact(rejector?: (value: any) => boolean): ArraySchema<T>;
 }
 
+export type ObjectSchemaDefinition<T extends object> = { [field in keyof T]: Schema<T[field]> | Ref };
+
 export interface ObjectSchemaConstructor {
-    <T>(fields?: { [field in keyof T]: Schema<T[field]> }): ObjectSchema<T>;
+    <T extends object>(fields?: ObjectSchemaDefinition<T>): ObjectSchema<T>;
     new (): ObjectSchema<{}>;
 }
 
-export interface ObjectSchema<T> extends Schema<T> {
-    shape(
-        fields: { [field in keyof T]: Schema<T[field]> },
+export interface ObjectSchema<T extends object> extends Schema<T> {
+    shape<U extends object>(
+        fields: ObjectSchemaDefinition<U>,
         noSortEdges?: Array<[string, string]>
-    ): ObjectSchema<T>;
+    ): ObjectSchema<T & U>;
```

Comment:
Does this have the same properties of overwriting the last property type? eg
```
object({
  a: string(),
  b: number(),
}).shape({
  b: string(),
  c: number(),
});
```
Should become:
```
object({
  a: string(),
  b: string(),
  c: number(),
});
```
where property b is overwritten, however, I don't think TS has the same semantics.

I did a quick test
```
interface A {
  a: string;
  b: string;
}
interface B {
  b: number;
  c: number;
}

type X = A & B;

const x: X = {
  a: 'asd',
  b: 23,
  c: 123,
};
```

and got this:
<img width="252" alt="screen shot 2018-12-12 at 11 41 29" src="https://user-images.githubusercontent.com/4065840/49864551-ff373d00-fe02-11e8-95f8-eb7d22df8667.png">

<img width="421" alt="screen shot 2018-12-12 at 11 41 41" src="https://user-images.githubusercontent.com/4065840/49864569-052d1e00-fe03-11e8-9e79-0cc59267d886.png">

Reply:
Good catch. I guess more research is needed here. In the end, we might have to pick the lesser of two evils.

Reply:
So it depends on how pedantic we want to be. According to the Typescript standard library, the return type of `Object.assign<T, U>(dest: T, source: U)` is `T & U` which is essentially what we're doing here but as @sseppola pointed out, the behavior can be different in Typescript. Going back to the conditional types PR from Anders, we can see that [he has a solution](https://github.com/Microsoft/TypeScript/pull/21316#issuecomment-359574388) in the form of a spread type which is equivalent to `{ ...dest, ...source }` that he calls `Spread<T, U>`. 

If we think that this makes sense to us, we can adopt it along with some comments that explain what we're doing because it's fairly advanced Typescript and we want to keep this approachable to newcomers.

Reply:
After reviewing the PR from Anders, the `Spread<T, U>` helper type is indeed advanced. It appears to me that it grows complicated because it is written for the fully generalized use case that must account merging optional properties correctly.

The good news is our use case is more specific--we only need to concern ourselves with merging schema definitions. They don't hold optional properties and for property names in common the type of last occurrence wins. For review, here's yup's documentation on `yup.object().shape()`.

> Note that you can chain shape method, which acts like object extends, for example:
> 
> ```javascript
> object({
>   a: string(),
>   b: number(),
> }).shape({
>   b: string(),
>   c: number(),
> });
> ```
> would be exactly the same as:
> 
> ```javascript
> object({
>   a: string(),
>   b: string(),
>   c: number(),
> });
> ```

Focusing on property `b`, merging `b: number()` with `b: string()` results in `b: string()` because it occurs last in the chain. So I think we can get away with `Shape<T, U>` (named for its specificity the `yup.object().shape()` method):

```typescript
type Shape<T extends object, U extends object> = { [P in keyof T]: P extends keyof U ? U[P] : T[P] } & U;
```

Applying this to the yup example above:

```typescript
interface AB {
    a: string;
    b: number;
}

interface BC {
    b: string;
    c: number;
}

interface ExpectedABC {
    a: string;
    b: string;
    c: number;
}

const abc: yup.Shape<AB, BC> = {
    a: 'qwerty',
    b: 'asdfg',
    c: 123
}; // $ExpectType ExpectedABC
```

That compiles just fine. And finally, putting it to use in `ObjectSchema.shape()` looks like this:

export interface ObjectSchema<T extends object> extends Schema<T> {
    shape<U extends object>(
        fields: ObjectSchemaDefinition<U>,
        noSortEdges?: Array<[string, string]>
    ): ObjectSchema<Shape<T, U>>;
```

How's that look to you guys? In the meantime, I'm going to update this PR assuming this will be the go forward solution.